pkg/controller/datavolume-conditions.go

```diff
@@ -0,0 +1,158 @@
+/*
+Copyright 2020 The CDI Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controller
+
+import (
+	"strings"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	cdiv1 "kubevirt.io/containerized-data-importer/pkg/apis/core/v1alpha1"
+)
+
+const (
+	boundFalse      = "BoundChangeFalse"
+	transferRunning = "TransferRunning"
+	pvcBound        = "PVCBound"
+	pvcPending      = "PVCPending"
+	claimLost       = "ClaimLost"
+	notFound        = "NotFound"
+)
+
+func findConditionByType(conditionType cdiv1.DataVolumeConditionType, conditions []cdiv1.DataVolumeCondition) *cdiv1.DataVolumeCondition {
+	for i, condition := range conditions {
+		if condition.Type == conditionType {
+			return &conditions[i]
+		}
+	}
+	return nil
+}
+
+func updateCondition(conditions []cdiv1.DataVolumeCondition, conditionType cdiv1.DataVolumeConditionType, status corev1.ConditionStatus, message, reason string) []cdiv1.DataVolumeCondition {
+	condition := findConditionByType(conditionType, conditions)
+	if condition == nil {
+		conditions = append(conditions, cdiv1.DataVolumeCondition{
+			Type: conditionType,
+		})
+		condition = findConditionByType(conditionType, conditions)
+	}
+	if condition.Status != status {
+		condition.LastTransitionTime = metav1.Now()
+	}
+	condition.Status = status
+	condition.Message = message
+	condition.Reason = reason
```

Comment:
I think we should also only update message/reason when the status changes

Reply:
I went back and forth on this, and decided on always updating because its possible to stay at the same condition but have a different message/reason, I thought it was important to always have the latest message/reason, but if I do that, I should have the last transition time also be dependent on reason/message being different.

Reply:
I think the original reason is more important and don't want to lose it